### ISSUE TITLE
[BR]: Insufficient patching for systemd/Debian

### ISSUE THREAD

**Issue by Artoria2e5** [fail2ban/fail2ban#3768]
On Debian 12 which uses systemd, fail2ban simply dies because `paths-debian.conf` points a legacy, nonexistent path. See https://superuser.com/questions/1830245/i-cant-get-fail2ban-working-on-debian-12.

There is already an upstream Debian patch for that: https://sources.debian.org/patches/fail2ban/1.1.0-4/update_backend_system.diff/.

Considering that Debian 12 has been the stable Debian version for 7 months, it's probably a good idea to pull that fix in.

**Comment by Artoria2e5** [comment]
So uh although the sshd part is fixed in d0d07285234871bad3dc0c359d0ec03365b6dddc, enabling anything else should cause some less-than-pleasant results. I think we should follow the Debian guys with a bigger patch.

**Comment by sebres** [comment]
> Considering that Debian 12 has been the stable Debian version for 7 months, it's probably a good idea to pull that fix in.

Probably.
@fail2ban/maintainers, @sylvestre what do you think about?

> I think we should follow the Debian guys with a bigger patch.

The problem with this is that the stock fail2ban is released by default with `paths-debian.conf` include, and many people never change it in their configurations (even if they are in completely different distro). So after new release it'd be totally incompatible.
One thing is to check backend by sshd jail only, another is to make that for a lot of jails.

However one can make a new `paths-whatever.conf` (like `paths-default.conf`) which would be backwards compatible and include it instead for stock release. But... it is again a large compat thing, because then local changes in `paths-debian.local` will not work after upgrade anymore.

**Comment by sebres** [comment]
As my forecast said, it is happening here - #3891 (after update a jail sees nothing, because default backend has become `systemd`, so monitoring journal, whereas the jail was intended to monitor file).

I'm really in mode to extend `auto` backend in the way, that it'd prefer file-related backends, as long as logpath exists, otherwise switches to the `systemd` (if journalmatch specified). Alternatively also with some other option allowing this behavior.

**Comment by sylvestre** [comment]
Sorry, i missed your ping

to be clear, you would like me to replace systemd by auto in
https://salsa.debian.org/python-team/packages/fail2ban/-/blob/master/debian/patches/update_backend_system.diff?ref_type=heads ?
correct?

**Comment by sebres** [comment]
No, at the moment this `auto` enhancement is not implemented yet (but I'd probably do it).
And the patch, you show above, sets `systemd` only to variables, so for example the change is only for those using interpolations `%(syslog_backend)s`  etc. What may be correct or not, but it'd affect only such jails.

The particular case of #3891 was that `backend` **in default section**... I think it was something like this:
```diff
  [DEFAULT]
+ backend = systemd
```
where previously (as still the case in stock `jail.conf`) this is defined as `auto`.
No Idea where exactly it is specified, but it looks that Ubuntu package got this somewhere in some include or override at some point, but I guess it could be - e634fa863e2d8035181e9e03476ae6dd56044fe6. However it was fixed later in [c03b1a832132f803](https://salsa.debian.org/python-team/packages/fail2ban/-/commit/c03b1a832132f8033a6c698650daba9c48e22c62) because of https://github.com/fail2ban/fail2ban/issues/3292#issuecomment-2078361360.

Also if I understand correctly, it is the intention of this issue here (change default backend). My point was - it may be too dangerous (and may affect all the users across the world with all own jails).

Anyway the "wrong" default looks to be fixed now in debian (current master on salsa.debian.org doesn't have it in `debian/debian-files/jail.d_defaults-debian.conf`), so further update for Ubuntu may get the fixed default backend, but... 

We have to do something with this issue, so I tend to extend the `auto` in the way I suggested above.

**Comment by sebres** [comment]
New branch [systemd-review](https://github.com/fail2ban/fail2ban/compare/16ae53e888ee0c4e83d5bf8314f40c78de0b8f06...sebres:fail2ban:systemd-review) (commit b2352f113eed5c54fb6c85e9e916b3b1ee02a854) implements inter alia the feature of automatic switch to backend `systemd` (for `backend = auto`), when:
- no files matching `logpath` found for this jail;
- no `systemd_if_nologs = false` is specified for the jail (`true` by default);
- option `journalmatch` is already set for the jail or its filter (otherwise it'd be too heavy to allow all auto-jails, even if they have never been foreseen for journal);
- option `skip_if_nologs` will be ignored if we could switch backend to `systemd`.
